sdk/core/azure-mgmt-core/azure/mgmt/core/policies/_base.py

```diff
@@ -0,0 +1,119 @@
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+# --------------------------------------------------------------------------
+import json
+import logging
+import re
+import time
+import uuid
+
+from azure.core.pipeline import PipelineContext, PipelineRequest
+from azure.core.pipeline.policies import HTTPPolicy
+from azure.core.pipeline.transport import HttpRequest
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ARMAutoResourceProviderRegistrationPolicy(HTTPPolicy):
+    """Auto register an ARM resource provider if not done yet.
+    """
+
+    def send(self, request):
+        # type: (PipelineRequest[HTTPRequestType], Any) -> PipelineResponse[HTTPRequestType, HTTPResponseType]
+        http_request = request.http_request
+        response = self.next.send(request)
+        if response.http_response.status_code == 409:
+            rp_name = self._check_rp_not_registered_err(response)
+            if rp_name:
+                url_prefix = self._extract_subscription_url(http_request.url)
+                if not self._register_rp(request, url_prefix, rp_name):
+                    return response
+                # Change the 'x-ms-client-request-id' otherwise the Azure endpoint
+                # just returns the same 409 payload without looking at the actual query
+                if "x-ms-client-request-id" in http_request.headers:
+                    http_request.headers["x-ms-client-request-id"] = str(uuid.uuid4())
+                response = self.next.send(request)
+        return response
+
+    @staticmethod
+    def _check_rp_not_registered_err(response):
+        try:
+            response = json.loads(response.http_response.text())
+            if response["error"]["code"] == "MissingSubscriptionRegistration":
+                match = re.match(r".*'(.*)'", response["error"]["message"])
+                return match.group(1)
+        except Exception:  # pylint: disable=broad-except
+            pass
+        return None
+
+    @staticmethod
+    def _extract_subscription_url(url):
+        """Extract the first part of the URL, just after subscription:
+        https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/
+        """
+        match = re.match(r".*/subscriptions/[a-f0-9-]+/", url, re.IGNORECASE)
+        if not match:
+            raise ValueError("Unable to extract subscription ID from URL")
+        return match.group(0)
+
+    @staticmethod
+    def _build_next_request(initial_request, method, url):
+        request = HttpRequest(method, url)
+        context = PipelineContext(
+            initial_request.context.transport, **initial_request.context.options
+        )
+        return PipelineRequest(request, context)
+
+    def _register_rp(self, initial_request, url_prefix, rp_name):
```

Comment:
Do you want to have timeout?

Reply:
We use the pipeline itself to send the request, and we duplicate the **kwargs from the main request (including timeout) so we should be fine